casper/src/main/scala/coop/rchain/casper/engine/ApproveBlockProtocol.scala

```diff
@@ -69,17 +72,32 @@ object ApproveBlockProtocol {
       interval: FiniteDuration
   ): F[ApproveBlockProtocol[F]] =
     for {
-      genesisBlock <- Genesis.fromInputFiles[F](
-                       maybeBondsPath,
-                       numValidators,
-                       genesisPath,
-                       maybeVaultsPath,
-                       minimumBond,
-                       maximumBond,
-                       shardId,
-                       deployTimestamp
+      now       <- Time[F].currentMillis
+      timestamp = deployTimestamp.getOrElse(now)
+
+      vaults <- VaultParser.parse[F](maybeVaultsPath, genesisPath.resolve("wallets.txt"))
+      bonds <- BondsParser.parse[F](
+                maybeBondsPath,
+                genesisPath.resolve("bonds.txt"),
+                numValidators,
+                genesisPath
+              )
+      validators = bonds.toSeq.map(Validator.tupled)
+      genesisBlock <- createGenesisBlock(
+                       implicitly[RuntimeManager[F]],
+                       Genesis(
+                         shardId = shardId,
+                         timestamp = timestamp,
+                         proofOfStake = ProofOfStake(
+                           minimumBond = minimumBond,
+                           maximumBond = maximumBond,
+                           validators = validators
+                         ),
+                         vaults = vaults,
+                         supply = Long.MaxValue
+                       )
```

Comment:
Why have you decided to inline it here? If there is a good reason for that, shouldn't you delete `Genesis.fromInputFiles` since it isn't used anywhere else? BTW, you have also lost the test coverage by inlining it here since `GenesisTest` no longer tests this piece of code.

Reply:
That function was doing too  much and I needed to split it into parsing the bonds file and using the bonds map. This way I had access to the bonds map so I can check that requiredSigs is smaller than the number of bonds.

The other problem is that the function was used in other places too so I had to inline it here and keep the other usages as they were.

Reply:
Leaving this inlined is a step back. Instead, you should separate the parsing logic and validation logic *within* `fromInputFiles` (ideally moving validation logic into `createGenesisBlcok`), and add the check in the "logic" part (it's valid for all use cases of Genesis.createGenesisBlock)

Reply:
That's what I did.I separated the parsing logic and moved bonds parser to util/BondsParser.scala and the vaults parser to util/VaultsParser.scala and then I called them instead of calling Genesis.fromInputFiles.

As for the validation logic I don't think it's a part of createGenesisBlock because the genesisBlock should not be aware of the requiredSigs parameter. That is specific to ApproveBlockProtocol so the validation should be done in its constructor i.e. the function ApproveBlockProtocol.of

Reply:
@odeac it is possible to access genesis bonds without actually having access to the parsed data structure from `BondsParser`. They are available through `ProtoUtil.bonds(genesis)` which internally just calls `genesis.body.get.state.get.bonds`. This way you don't have to touch `Genesis.fromInputFiles` at all.

Reply:
That seems to be a good solution. I'll do that

Reply:
Correction: I've opened a discussion in #blockchain to see if we keep that field or if it should be removed. I'll continue with this PR after we clarify this aspect

Reply:
> As for the validation logic I don't think it's a part of createGenesisBlock because the genesisBlock should not be aware of the requiredSigs parameter. That is specific to ApproveBlockProtocol so the validation should be done in its constructor i.e. the function ApproveBlockProtocol.of

It is an integrity check for our config values, regardless of where it's used. `bonds.size == validators.size`, and `validators` is part of `Genesis`. Hence, the smart constructor for Genesis should validate that constraint. All the parties in the genesis (including ceremony master) should have same config values for requiredSigs and initial bonds.

Reply:
@itegulov Following the discussion sin #blockchain discord channel, the decision is that we should remove BlockMessage.body.state.bonds so this is no longer a valid solution

Reply:
@ArturGajowy So you want to add a parameter named requiredSigs in createGenesisBlock such that I can verify the condition that requiredSigs < bonds.size and fail if that's not the case?

Reply:
I thought `requiredSigs` belongs to the `Genesis` case class, so I'd add a field there if it's not there already.

Reply:
> I'd add a field there if it's not there already.

No, that's wrong. You can actually validate the `requiredSigs >= bondedValidators` by having `val bondedValidators = genesis.proofOfStake.validators.size`

Reply:
The function Genesis.fromInputFiles doesn't return an instance of the Genesis case class. It returns a BlockMessage representing the genesis block. That's the problem.

I don't understand why do we care about the Genesis.fromInputFiles function which is only used in one place in the tests and one place in production. I've inlined both its uses and we don't have any problems.

Reply:
> I've inlined both its uses and we don't have any problems.

I haven't noticed that - it changed during the works, as you've mentioned. What you've done with moving out the config parsing functions makes sense :+1: